[RFC 0002] Flux OCI support for Helm (#690)
fluxcd/source-controller: docs/spec/v1beta2/helmrepositories.md

# Helm Repositories

> There are 2 Helm repository types defined by the `HelmRepository` API:

Reviewer: ```suggestion
There are 2 [Helm repository types](#types) defined by the `HelmRepository` API:
```

Should have suggested this yesterday already, sorry.

> There are 2 Helm repository types defines by the `HelmRepository` API: 

Reviewer: I would change the order of the two entries, this aligns with the order below and follows historical order.

A HelmRepository also needs a
[`.spec` section](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#spec-and-status).

> The HelmRepository API defines a field `spec.type` to specify a Helm repository type. 

Reviewer: This should be a separate heading.

```
### Type

`.spec.type` is a [..]
```

A HelmRepository also needs a
[`.spec` section](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#spec-and-status).

### Type

`.spec.type` is an optional field that specifies the Helm repository type. 

> Possible values are `default` for a Helm HTTP/S repository or `oci` for an OCI Helm repository.

Reviewer: ```suggestion
Possible values are `default` for a Helm HTTP/S repository, or `oci` for an OCI Helm repository.
```

### URL

> `.spec.url` is a required field that specifies:
> - the HTTP/S address of the Helm repository for HTTP/S Helm repositories,
> - a valid OCI registry URL without the HelmChart's name for OCI Helm repositories 
> i.e. `oci://<registry>/<repository>`.

Reviewer: ```suggestion
`.spec.url` is a required field that specifies the HTTP/S or OCI address of a Helm repository depending on the [type of the HelmRepository object](#type). For OCI, the URL is expected to point to an OCI registry repository, i.e. `oci://ghcr.io/fluxcd/source-controller`.
```

maybe?

### URL

> `.spec.url` is a required field that specifies the HTTP/S or OCI address of a
> Helm repository depending on the [type of the HelmRepository object](#type).

Reviewer: ```suggestion
`.spec.url` is a required field that depending on the [type of the HelmRepository object](#type)
specifies the HTTP/S or OCI address of a Helm repository. 
```

It was late yesterday night, but looking at it now, I think this order would work better (and also: is better English).

  password: 123456
```

OCI Helm repository example:

```yaml
---
apiVersion: source.toolkit.fluxcd.io/v1beta2
kind: HelmRepository
metadata:
  name: podinfo
  namespace: default
spec:
  interval: 5m0s
  url: oci://ghcr.io/stefanprodan/charts
  type: "oci"
  secretRef:
    name: oci-creds
---
apiVersion: v1
kind: Secret
metadata:
  name: oci-creds
  namespace: default
stringData:
  username: example
  password: 123456
```

#### TLS authentication

> > **_NOTE:_**  TLs authentication is not yet supported by OCI Helm repositories.

Reviewer: ```suggestion
**Note:**  TLS authentication is not yet supported by OCI Helm repositories.
```

to match styling of already existing note entries.

provide better (timeout) support to solutions polling the HelmRepository to become
`Ready`.

The `artifact` and `storage` related condition types are only supported by
HTTP/s Helm repositories. OCI Helm repositories do not use storage, they do not have

> any artifacts.

Reviewer: I would instead maybe simply state that the OCI Helm repositories only make use of Reconciling, Ready and Stalled. Which is easier to comprehend than ruling out what it doesn't.

[typical-status-properties]: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties
[kstatus-spec]: https://github.com/kubernetes-sigs/cli-utils/tree/master/pkg/kstatus

> ## OCI Hem Repositories

Reviewer 2: I think we need to link this from the top level section and redo all headers. After the top level title, we could link to:
- HTTP/S Helm Repositories
- OCI Helm Repositories

@hiddeco may have a different idea.

Reviewer: As I went through great lengths to not duplicate any information in this document, I feel that this could indeed be better structured. My proposal would be the following:

- Delete this whole new section.
- Change introduction of document ("[..] API defines a Source to produce an Artifact [..]") to highlight there are two types.
- Add sub-sections to "examples":

	1. Example Helm HTTP/S repository
	2. Example Helm OCI repository

  Where the OCI section contains the same outline of information around the reconciler as what currently exists for HTTP/S.
- Add section for "Type" to "Writing a HelmRepository spec"
- Add note to "TLS authentication" that it only applies to non-OCI.
- Add a note about image pull secrets for OCI being worked on in https://github.com/fluxcd/flux2/pull/2601
- Instead of duplicating the Conditions section, find a way to highlight that the OCI type just has a `Ready` condition and no artifact.

Reviewer 3: Thanks for this! I have updated the doc. I did not add the part regarding the image pull secret though, I felt like it is better to update the doc when we implement it.
